### PR TITLE
Command-line interface

### PR DESCRIPTION
**Pull request by weijiekoh**
This PR implements a command-line interface that makes it possible to deploy and interact with an on-chain instance of MACI.

I developed a CLI instead of a web UI for practical reasons. To create a web UI has a lower return on time and effort invested, and a CLI does the job just as well despite the appearance.

After this PR, I will implement more comprehensive integration tests which involve more users, multiple batches, and bribery attacks.

**Subcommands**

The CLI has one command with several commands: `genMaciKeypair`, `genMaciPubkey`, `create`, `signup`, `publish`, `process`, and `tally`.

Refer to the [CLI readme](https://github.com/weijiekoh/maci/blob/feat/cli/cli/README.md) for details about each subcommand.

**Example flow**

1. Coordinator generates their MACI keypair and funds an Ethereum externally owned account with some ETH.
2. Some users do the same.
3. Coordinator runs `create` to deploy a new MACI contract, and shares the contract address with the users.
4. Users run `signup` to add their MACI public keys to the contract.
5. Users run `publish` to cast votes and/or change their keys.
6. Coordinator runs `process` to process all batches of messages.
7. Coordinator runs `tally` to tally all batches of state leaves.

**Running subcommands**

1. Build the codebase and download the zk-SNARK files following the directions in the main readme file.
2. In the `cli` subdirectory, run:

```bash
node ./build/index.js <subcommand> <options>
```

The `process` and `tally` subcommands require you to specify the `--max-old-space` Node option as they each generate zk-SNARK proofs, which is memory-intensive:

```bash
NODE_OPTIONS=--max-old-space-size=4096 node ./build/index.js <subcommand> <options>
```

**MACI keypairs**

A MACI keypair is an EdDSA public and private key.

The various CLI subcommands requires the coordinator and user to specify public and/or private keys as _serialised_ MACI private keys. These look like:

```
Private key: macisk.e11e7e8870b6b3472e04a1ed703306ffbc9fd4ae78eff0c8b6cb552d467ef3f
Public key:  macipk.483e37f5f4a11713d1a9c9df79fc7195f0e1e303e1d3bb823822e1930e1b2aa7
```

A MACI public key is the hexadecimal representation of the result of `babyJub.packPoint()` on itself.

For users' convenience, the `genMaciKeypair` subcommand supports a `--passphrase` option which accepts a passphrase (min. 32 characters) and runs a deterministic key-stretching algorithm on it to generate a MACI private key and its associated public key. The algorithm is Argon2id with a time cost of 30, memory cost of 40960, and salt `0x27864203c107e4b8c39e80eca12b7637` (just a randomly generated value with no real-world significance).

**Limitations**

The state tree depth, message tree depth, and vote option tree depth, message batch size, and tally batch sizes are currently all hardcoded to 4. In the future, we need to decide on the set of combinations of values which users may specify for these options, keeping in mind that each additional combination entails a new circuit and therefore requires a new trusted setup.

**Other modifications**

I also modified `batchUpdateStateTree.circom` to include a `msg_tree_batch_end_index` public input. This allows the coordinator to process all messages even if the total number of batches is not a perfect multiple of the message batch size.

### CODE REVIEW

File: cli/README.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+This command deploys an instance of a MACI contract.
+
+Fields that the coordinator has to set:
+
+`maci-cli create <options>`
+
+Example usage:
+
+```
+$ node build/index.js create -sk macisk.23d007423d56475d7e39dcd5053c5aa98f57a69ee85bc7813ccbf4c5e688307  -d 0xc87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3 -u 32768 -m 32768
```
Comment (ChihChengLiang) [inline cli/README.md:42]: Should we remove the `-u 32768 -m 32768` here as it currently not supported?

File: cli/README.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+Examples of serialized public and private keys:
+
+```
+Private key: macisk.e11e7e8870b6b3472e04a1ed703306ffbc9fd4ae78eff0c8b6cb552d467ef3f
+Public key:  macipk.483e37f5f4a11713d1a9c9df79fc7195f0e1e303e1d3bb823822e1930e1b2aa7
+```
+
+### Coordinator: Create election
+
```
Comment (ChihChengLiang) [inline cli/README.md:32]: What do you think adding `cd ../contracts && npm run ganache`?

File: cli/README.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+| Maximum number of vote options | `-v` or `--max-vote-options` | Default: 3 |
+| Sign-up duration | `-s` or `--signup-duration` | The sign-up duration, in seconds. Default: 3600. |
+| Voting duration | `-o` or `--voting-duration` | The voting duration, in seconds. Default: 3600. |
+| Initial voice credits | `-c` or `--initial-voice-credits` | Default: 100 |
+| Initial voice credit proxy contract | `-i` or `--initial-vc-proxy` | If specified, deploys the MACI contract with this address as the initial voice credit proxy constructor argument. Otherwise, deploys a ConstantInitialVoiceCreditProxy contract with the above-specified value. |
+| Signup gatekeeper contract | `-g` or `--signup-gatekeeper` | If specified, deploys the MACI contract with this address as the signup gatekeeper constructor argument. Otherwise, deploys a gatekeeper contract which allows any address to sign up. | 
+| Batch size for processing messages | `-bm` or `--message-batch-size` | Default: 4 |
+| Batch size for tallying votes | `-bv` or `--tally-batch-size` | Default: 4 |
+
+### Process, tally and prove outcome
```
Comment (ChihChengLiang) [inline cli/README.md:69]:
```suggestion
### Coordinator: Process, tally and prove outcome
```

File: contracts/scripts/runGanache.sh [hunk 4]
```diff
@@ -21,3 +21,6 @@ npx etherlime ganache --mnemonic "candy maple cake sugar pudding cream honey ric
 #7: 0x0f4f2ac550a1b4e2280d04c21cea7ebd822934b5
 #8: 0x6330a553fc93768f612722bb8c2ec78ac90b3bbc
 #9: 0x5aeda56215b167893e80b4fe645ba6d5bab767de
+
+# Private keys
```
Comment (ChihChengLiang) [inline contracts/scripts/runGanache.sh:25]: Is this accidentally checked in?

Comment (weijiekoh) [reply contracts/scripts/runGanache.sh:25]: Fortunately not!

This mnemonic used to be the default Truffle suite mnemonic. https://www.trufflesuite.com/docs/truffle/getting-started/using-truffle-develop-and-the-console

As such these private keys were compromised a long time ago.

File: cli/README.md (hunk 2 above)
Comment (weijiekoh) [reply cli/README.md:32]: Done :)

File: cli/README.md (hunk 1 above)
Comment (weijiekoh) [reply cli/README.md]: Done :)

File: circuits/circom/updateStateTree.circom [hunk 5]
```diff
@@ -35,10 +32,10 @@ template UpdateStateTree(
     // Output: New state tree root
     signal output root;
 
-    // Input(s)
+    // The coordinator's public key
     signal input coordinator_public_key[2];
 
-    // Note: a message is an encrypted command
+    // Note that a message is an encrypted command
     var message_length = 11;
```
Comment (ChihChengLiang) [inline circuits/circom/updateStateTree.circom:39]: nitpick: Should we capitalize `message_length`, `message_signature_length`, and `message_without_signature_length` too?

File: contracts/ts/deploy.ts [hunk 6]
```diff
@@ -64,48 +85,78 @@ const deploySignupTokenGatekeeper = async (
...
     signUpGatekeeperAddress: string,
     initialVoiceCreditProxy: string,
+    stateTreeDepth: number = config.maci.merkleTrees.stateTreeDepth,
+    messageTreeDepth: number = config.maci.merkleTrees.messageTreeDepth,
+    voteOptionTreeDepth: number = config.maci.merkleTrees.voteOptionTreeDepth,
+    quadVoteTallyBatchSize: number = config.maci.quadVoteTallyBatchSize,
+    messageBatchSize: number = config.maci.messageBatchSize,
+    voteOptionsMaxLeafIndex: number = config.maci.voteOptionsMaxLeafIndex,
+    signUpDurationInSeconds: number = config.maci.signUpDurationInSeconds,
+    votingDurationInSeconds: number = config.maci.votingDurationInSeconds,
```
Comment (ChihChengLiang) [inline contracts/ts/deploy.ts:114]: Looks like worth defining an interface for these configs.

File: circuits/circom/updateStateTree.circom (hunk 5 above)
Comment (weijiekoh) [reply circuits/circom/updateStateTree.circom:39]: Absolutely! Thanks for spotting this!

**Pull request merged**